Make name compression be based on CBOR-packed (#7)
cbor-wg/cbor-dns: draft-lenders-dns-cbor.md

~~~
{:cddl #fig:dns-response title="DNS Response Definition"}

# Compression with CBOR-packed {#sec:cbor-packed}

If both DNS server and client support CBOR-packed {{-cbor-packed}}, it MAY be used for further
compression in DNS responses.
Especially IPv6 addresses, e.g., in AAAA resource records can benefit from straight referencing to
compress common address prefixes.[^6]{: mlenders}

[^6]: Needs rewording.

## Name Compression {#sec:name-compression}

~~~ cddl
Text-String-Suffix-Sequence-Packed-CBOR = #6.28259(rump)
~~~

For name compression, a new packing table setup tag TBD28259 ('n' and 'c' in ASCII) for CBOR-packed {{-cbor-packed}} is defined.

> It provides an implicit text string suffix sequence table for shared items _V_ which is prepended to the existing table for shared items that apply to the entire tag TBD28259 (by default empty table).

Reviewer: ```suggestion
It provides an implicit text string suffix sequence table for shared items _V_ which is prepended to the existing table for shared items that apply to the content of tag TBD28259 (by default empty table).
```
?

> The this implicit (i.e. not explicitly represented) table _V_ is constructed as follows:

Reviewer: ```suggestion
This implicit (i.e. not explicitly represented) table _V_ is constructed as follows:
```

> Any coherent sequence of text strings encountered within the rump of tag TBD28259, as well as any of its non-empty suffixes, are added to the table as an array in depth-first order.

Reviewer: ```suggestion
Any coherent sequence of text strings encountered within the rump of tag TBD28259, as well as any of its non-empty suffixes, are added to the table as arrays in depth-first order.
```

Reviewer: Also, what exactly does "coherent sequence" mean? Is that a standing term?

Author: „CBOR sequence“ is a standing term. Cohererent is meant wrt text strings, so a (sub) sequence that only contains text strings.

Reviewer: Then maybe "any CBOR sequence consisting of only text strings"? Maybe it's also just a personal English-word problem with "coherent".

Text string sequences within any tables for shared items or argument items within the rump MUST not be added to _V_.
If a sequence for which an array is already in _V_ is encountered, a shared item reference _i_ to that array in V replaces this sequence.
This shared item reference _i_ means: take the sequence from the array at _V_\[_i_\] and put it into the surrounding array in place of _i_.
The resulting rump should look like referencing the _i_-th string (depth first) in the sequence.

The "application/dns+cbor" media type comes with an optional parameter "packed".
If it is not provided, it is assumed to be 0.

> With packed=0, any CBOR object `obj` marked by the "application/dns+cbor" media type MUST explicitly be understood as `TBD28259(obj)`, unless it is already `obj` itself is already tagged explicitly with TBD28259 as a whole.

Reviewer: I personally find it confusing that `packed=0` still means there is packing involved. Could we find a different naming?

### Example

> Take the following content (note that this is intentionally not legal "application/dns+cbor" to illustrate generality).

Reviewer: ```suggestion
Take the following CBOR object _o_ (note that this is intentionally not legal "application/dns+cbor" to illustrate generality).
```

or any other name, just to be able to reference it later.

~~~ edn
[
  "www", "example", "org",
  ["svc", "www", "example", "org"],
  "org", "example", "org", 42,
  "svc", "www", "example", "org", 42
]
~~~
{: #fig:name-compression-example-unpacked title="Unpacked example for implicit text string suffix sequence compression."}

This would generate the following virtual table _V_.

~~~ edn
[
    ["www", "example", "org"],
    ["example", "org"],
    ["org"],

>     ["svc", "www", "example", "org"],

Reviewer: ```suggestion
    ["svc", simple(0)],
```

is it not?

Author: Since it is implicit, this is arbitrary. I thought it might be more helpful to the reader, if we show the full sequences here.

Reviewer: I think this illustrates the implicit table setup algorithm better.

    ["org", "example", "org"]
]
~~~
{: #fig:name-compression-example-table title="Implicit table of shared items for the example."}

> Note that the sequence "org", "example", "org" is added at index 4 with leading "org", instead of referencing index 2 + index 1, as it is its own distinct suffix sequence.

Reviewer: ```suggestion
Note that the sequence "org", "example", "org" is added at index 4 with leading "org", instead of referencing index 2 + index 1 (`simple(2), simple(1)`), as it is its own distinct suffix sequence.
```

> The packed representation would thus be:

Reviewer: ```suggestion
The packed representation of _o_ would thus be:
```

~~~ edn
TBD28259(
  [
    ["www", "example", "org"],
    ["svc", simple(0)],
    "org", simple(1), 42,
    simple(3), 42
  ]
)
~~~
{: #fig:name-compression-example-packed title="The packed representation of the example."}

Note, with "application/dns+cbor;packed=0" the surrounding TBD28259 can be elided (even though the content would not be parsable as application/dns+cbor).

> With, e.g., table setup tag 113, further packing can be achieved by nesting table packing them with TBD28259.

Reviewer: ```suggestion
With, e.g., table setup tag 113, further packing can be achieved via nesting table packing.
```

~~~ edn

> TBD28259(
>   113(

Reviewer: ```suggestion
113(
  TBD28259(
```

otherwise prepending the table from 113 would shift the TBD28259 ones.

Reviewer: Which clashes with the requirement that we want to elude TBD28259 for dns+cbor, hum...

Author: > otherwise prepending the table from 113 would shift the TBD28259 ones.

Huh? From how I read my definition (and the definition for 113), the tables of the outer tag are prepended to the tables of the inner tag. I want the TBD28259 table _V_ to be the “start” of the shared item table.

Reviewer: At least for 113, it is definitely the other way around. I guess the rationale is that you enable shorter references within smaller contexts (inside the rump of an inner tag).

Author: Are you sure? Maybe @cabo can help to alleviate the confusion? Does the outer setup tag prepend its table(s) to the inner tag's table(s) or the inner tag to its outer tag's?

Reviewer 2: The main argument is really composability.  A CBOR data item that uses cbor-packed should still work when it is embedded in some other table setup.

Author: Ok, so in this case, if I understand correctly, the outer tag should append to the packing table(s) of the inner tag.

E.g., taking the example

```
113(
  TBD28259(
    [
      ["org", 42],
      [
        ["www", "example", simple(5)],
        ["svc", simple(0)],
        simple(5), simple(1), simple(6),
        simple(3), simple(6)
      ]
    ]
  )
)
```

would still work with

```
  TBD28259(
    [
        ["www", "example", "org"],
        ["svc", simple(0)],
        "org", simple(1), 42,
        simple(3), 42
    ]
  )
```

while the other way around 

```
TBD28259(
  113(
    [
      ["org", 42],
      [
        ["www", "example", simple(5)],
        ["svc", simple(0)],
        simple(5), simple(1), simple(6),
        simple(3), simple(6)
      ]
    ]
  )
)
```

would completely need to rewrite the `simple(i)` references (and not just the new ones) for `113(...)` if `TBD28259(...)` is taken away to make sense.

```
  113(
    [
      ["org", 42],
      [
        ["www", "example", simple(0)],
        ["svc", "www", "example", simple(0)],
        simple(0), "example", simple(0), sample(1),
        "svc", "www", "example", "org", sample(1)
      ]
    ]
  )
```

This makes sense to me. Will fix (and try to find a better wording to make the table setup clearer).

Reviewer 2: > Ok, so in this case, if I understand correctly, the outer tag should append to the packing table(s) of the inner tag.

That is a funny way to say "prepend tables built from inner tags to those inherited from an outer table building tag", but it definitely is what has worked well so far.
(An on-the-fly implementation could start at the node with the reference, go up until it finds the next table building tag, use what it finds there if the reference is covered, or subtract the size of the table from a reference index that goes beyond that table and continue upwards.)

    [
      ["org", 42],
      [
        ["www", "example", simple(5)],
        ["svc", simple(0)],
        simple(5), simple(1), sample(6),
        simple(3), sample(6)
      ]
    ]
  )
)
~~~
{: #fig:name-compression-example-packed-113 title="The packed representation of the example with additional table setup."}

> Note, how the previous references in {{fig:name-compression-example-packed}} do not changed, even though the string `"org"` is prepended, as strings in tables are ignored when creating the implicit table.

Reviewer: Could we get rid of this extra rule with by switching the tag order?

Author: Then the indices for name compression would be larger than the 113 ones... Which would increase the risk of names having larger references (but as they are more common in DNS messages they should be smaller).

Reviewer: see above.

~~~

For name compression, a new packing table setup tag TBD28259 ('n' and 'c' in ASCII) for CBOR-packed {{-cbor-packed}} is defined.
It provides an implicit text string suffix sequence table for shared items _V_ which is appended to the existing table for shared items of any table setup tag within the content of tag TBD28259 (by default empty table).
This implicit (i.e. not explicitly represented) table _V_ is constructed as follows:
Any coherent sequence of text strings encountered within the rump of tag TBD28259, as well as any of its non-empty suffixes, are added to the table as arrays in depth-first order.
Text string sequences within any tables for shared items or argument items within the rump MUST not be added to _V_.
If a sequence for which an array is already in _V_ is encountered, a shared item reference _i_ to that array in V replaces this sequence.
This shared item reference _i_ means: take the sequence from the array at _V_\[_i_\] and put it into the surrounding array in place of _i_.
The resulting rump should look like referencing the _i_-th string (depth first) in the sequence.

The "application/dns+cbor" media type comes with an optional parameter "packed".

> If it is not provided, the value of it, i.e. the level of packedness, is assumed to be 0.

Author: I am not set on the term “level of packedness”. In fact, I would love to get some alternative terms.

In the registry "{{cbor-tags (CBOR Tags)<IANA.cbor-tags}}" {{IANA.cbor-tags}},
IANA is requested to allocate the tags defined in {{tab-tag-values}}.

|    Tag   | Data Item | Semantics                                                            | Reference              |
| TBD141   | array     | CBOR EDNS option record                                              | draft-lenders-dns-cbor |

> | TBD28259 | any       | Packed CBOR; implicit text string suffix sequence shared-item table  | draft-lenders-dns-cbor |

Reviewer: everything except `tag`, `array` and `map` wouldn't make sense, right?

Author: That would be the same for 113 or 1113. Nevertheless, they have any as data item as well.
